[ML] Do not tie features to product versions.

edsavage opened the pull request:
Move away from interrogating the version of the node(s) in order to determine if they are suitable to support particular features of the MachineLearning plugin. Instead make use of a new `MlConfigVersion` class, that is backwardly compatible with the product versioning scheme, but going forward can increment independently.

Labelling as `>non-issue` as this change is purely internal mechanics, unrelated to any new or existing feature.

edsavage committed:
[ML] Do not tie features to product versions.

Move away from interrogating the version of the node(s) in order to determine if they are suitable to support particular features of the MachineLearning plugin. Instead make use of a new `MlConfigVersion` class, that is backwardly compatible with the product versioning scheme, but going forward can increment indepently.

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

jonathan-buttner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,402 @@
+        return new Tuple<>(minMlConfigVersion, maxMlConfigVersion);
+    }
+
+    public static MlConfigVersion fromString(String str) {
```
I don't know how feasible it is to get a negative version but should we use `Integer.parseUInt` or check to ensure that the resulting number isn't negative?

edsavage replied:
Thanks Jonathan. I think the main issue here is that `MlConfigVersion` needs to be backwardly compatible with `Version` and hence it makes more sense to palm off the string parsing to `Version.fromString`

jonathan-buttner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,402 @@
+/**
+ * The version number associated with various ML features.
+ * <p>
+ * Prior to 8.18.0, the release {@link Version} was used everywhere. This class separates the ML config format version
```
Should this say `Prior to 8.10.0...`?

jonathan-buttner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,402 @@
+        return new Tuple<>(minMlConfigVersion, maxMlConfigVersion);
+    }
+
+    public static MlConfigVersion fromString(String str) {
```
Just to confirm, we don't need to handle the full version format defined by `Version`: https://github.com/edsavage/elasticsearch/blob/1d8b286f153f1339d7f6b89b2c4c5494372e87f7/build-tools/src/main/java/org/elasticsearch/gradle/Version.java#L66-L81 because the legacy versions we've registered above don't include things like `alpha` or `-SNAPSHOT`?

edsavage replied:
As above.. I think it makes sense to use the `Version.fromString` parser here, so yes, it would understand release candidates etc.

edsavage committed:
Attend to review comments.

Delegate string parsing to Version.fromString

edsavage committed:
Fix some integration test failures

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+import java.util.TreeMap;
+
+/**
+ * The version number associated with various ML features.
```
This comment needs to explain why this class is needed in addition to `TransportVersion`. It's because transport version cannot be persisted in stored documents or cluster state metadata. So this class is designed to be persisted in human-readable format, and indicate the age of that state or config.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+ * <h2>Version compatibility</h2>
+ * The earliest compatible version is hardcoded in the {@link #MINIMUM_COMPATIBLE} field. This cannot be dynamically calculated
+ * from the major/minor versions of {@link Version}, because {@code MlConfigVersion} does not have separate major/minor version numbers.
+ * So the minimum compatible version is hard-coded as the ML config version used by the first version of the previous major release.
+ * {@link #MINIMUM_COMPATIBLE} should be updated appropriately whenever a major release happens.
```
This part isn't true. We don't aggressively upgrade every single config when moving to a new major version. It's possible that there are some ML config documents in 8.x clusters that predate 7.0.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+    /**
+     * Returns {@code true} if the specified version is compatible with this running version of Elasticsearch.
+     */
+    public static boolean isCompatible(MlConfigVersion version) {
+        return version.onOrAfter(MINIMUM_COMPATIBLE);
+    }
```
This should be removed. It's not _necessarily_ the case that every ML config created prior to 7.0 is invalid now. The decision needs to be taken on a per-config basis by the code that deals with that config.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+        MlConfigVersion minMlConfigVersion = MlConfigVersion.CURRENT;
+        MlConfigVersion maxMlConfigVersion = MlConfigVersion.MINIMUM_COMPATIBLE;
+        for (DiscoveryNode node : nodes) {
+            String mlConfigVersionStr = node.getAttributes().get(mlConfigVersionNodeAttrName);
```
For nodes that don't have the attribute we should call `node.getVersion()` and infer the ML config version from that.

As this code stands it will convert a `null` attribute to the current version. This is guaranteed to be wrong. The version of the node that didn't have the attribute must be old.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MachineLearningField.java:
```diff
@@ -47,10 +46,10 @@ public final class MachineLearningField {
     // Ideally this would be 7.0.0, but it has to be 6.4.0 because due to an oversight it's impossible
     // for the Java code to distinguish the model states for versions 6.4.0 to 7.9.3 inclusive.
-    public static final Version MIN_CHECKED_SUPPORTED_SNAPSHOT_VERSION = Version.fromString("6.4.0");
+    public static final MlConfigVersion MIN_CHECKED_SUPPORTED_SNAPSHOT_VERSION = MlConfigVersion.fromString("6.4.0");
```
This is a great example of why `MINIMUM_COMPATIBLE` should not be 7.0.0.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:
```diff
@@ -806,6 +811,7 @@ public Settings additionalSettings() {
                 Integer.toString(EsExecutors.allocatedProcessors(settings))
             );
             addMlNodeAttribute(additionalSettings, allocatedProcessorsAttrName, Double.toString(getAllocatedProcessors().count()));
+            addMlNodeAttribute(additionalSettings, mlConfigVersionAttrName, MlConfigVersion.CURRENT.toString());
```
We're going to need this attribute on more than just ML nodes. Certainly master and ingest nodes will need it, as some ML operations are performed on master nodes and ingest nodes. But I'm thinking it's best to just add this attribute on all nodes.

The code that determines the min and max versions present in the cluster assumes that the attribute will be set on all nodes.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TrainedModelValidator.java:
```diff
@@ -56,7 +56,7 @@ static void validateMinimumVersion(ModelPackageConfig resolvedModelPackageConfig
         }
 
-        if (state.nodes().getMinNodeVersion().before(minimumVersion)) {
+        if (MlConfigVersion.fromVersion(state.nodes().getMinNodeVersion()).before(minimumVersion)) {
```
This should be using `MlConfigVersion.getMinMlConfigVersion()`.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java:
```diff
@@ -165,7 +166,8 @@ private static boolean nodeSupportsModelSnapshotVersion(DiscoveryNode node, Job
             return true;
         }
-        return node.getVersion().onOrAfter(job.getModelSnapshotMinVersion());
+        return MlConfigVersion.fromString(node.getAttributes().get(MachineLearning.ML_CONFIG_VERSION_NODE_ATTR))
+            .onOrAfter(job.getModelSnapshotMinVersion());
```
Instead of `MlConfigVersion.fromString(node.getAttributes().get(MachineLearning.ML_CONFIG_VERSION_NODE_ATTR))` here you should write a helper method in `MlConfigVersion`. Something like `getMlConfigVersionForNode(DiscoveryNode, String)`. That can then get the attribute and convert it to an `MlConfigVersion`, but also account for the case of a node that's too old to have the attribute (by looking at its `Version`).

In fact, that helper method could then be used in the loop in `getMinMaxMlConfigVersion`.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/task/OpenJobPersistentTasksExecutor.java:
```diff
@@ -182,7 +184,8 @@ public static String nodeFilter(DiscoveryNode node, Job job) {
         }
 
-        if (Job.getCompatibleJobTypes(node.getVersion()).contains(job.getJobType()) == false) {
+        if (Job.getCompatibleJobTypes(MlConfigVersion.fromString(node.getAttributes().get(MachineLearning.ML_CONFIG_VERSION_NODE_ATTR)))
```
As above, `MlConfigVersion.fromString(node.getAttributes().get(MachineLearning.ML_CONFIG_VERSION_NODE_ATTR))` should be replaced with a call to a new helper method.

And the same for anywhere else where this pattern is needed.

droberts195 reviewed (commented):
This looks like it's on the right track.

I made a quick first pass and came up with a number of suggestions. Please have a look at those and then I'll have a closer look.

edsavage committed:
Attend to review comments

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into ml_config_version

thecoop commented:
You might be interested in https://github.com/elastic/elasticsearch/pull/97506, which introduces a `VersionId` interface that we can pull a lot of the duplicated methods into - this'll help with a lot of the boilerplate (when that gets merged)

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/ElasticsearchMappings.java:
```diff
@@ -100,7 +101,7 @@ public class ElasticsearchMappings {
     private ElasticsearchMappings() {}
 
-    static String[] mappingRequiresUpdate(ClusterState state, String[] concreteIndices, Version minVersion) {
+    static String[] mappingRequiresUpdate(ClusterState state, String[] concreteIndices, MlConfigVersion minVersion) {
```
This function is used with ML system indices as well as hidden indices. System index versioning needs to be understood by core server code. So I don't think we can use `MlConfigVersion` for versioning our index mappings. We'll have to use the approach of https://github.com/elastic/elasticsearch/pull/97529.

There's more detail in https://github.com/elastic/elasticsearch/pull/97529#discussion_r1268181542 and my reply to that.

I think it's best if you change this PR to leave index mapping versioning as it was, and just change our config and state versioning. Sorry for the extra work.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+     * Reference to the earliest compatible ML config version to this version of the codebase.
+     * This is hard-coded as the ML config version used by the first version of the previous major release..
+     */
+    public static final MlConfigVersion MINIMUM_COMPATIBLE = V_6_0_0;
```
I think for ML this needs to be 5.4.0, when ML was first released to the world. Although we wouldn't actually be compatible with that now, it's possible that config docs could still be lying around in some customers' indices that reference that version. For clarity the constant should also be renamed, to something like `FIRST_ML_VERSION`.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+
+    @Override
+    public String toString() {
+        if (id >= MINIMUM_COMPATIBLE.id) {
```
I found this strange. It's a bit weird because we're using lower numbers for newer versions.

At the very least this deserves a comment here.

But I also wonder if the lack of comparability of the numeric IDs is going to bite us at some point. I think it would be better to use numeric IDs that are 10000000 greater than the printed number we'll use. Then the numeric IDs would compare naturally. In other words print 10000010 as "10".

thecoop replied:
We've taken the approach of just keeping the same number going - so our detached versions start at 8_500_000 and go up from there. Then it's just comparing a single number

droberts195 replied:
For usage in config documents we have the additional requirement to be able to serialize to and from text.

This is why I was thinking that it would be nice to use smaller numbers for the printed version. So we'd progress 8.9.0, 8.9.1, 8.9.2, 10, 11, 12, 13, .... But then it would also be nice if the internal single number ID was incrementing too.

However, just thinking about this a bit more, we'll still need to print in x.y.z format to support rolling upgrades.  If we print "10" into a config document and an 8.9 node reads that then it will throw an exception: https://github.com/elastic/elasticsearch/blob/568b292bdeb04103cb7d30f191a6378bac8f57ed/server/src/main/java/org/elasticsearch/Version.java#L255-L259

That would happen somewhere like here: https://github.com/elastic/elasticsearch/blob/4c8847bdd4c31aa2ddb04b0d5eb7c2968696a316/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java#L874

So I am actually thinking the best way to print these numbers is with .0.0 appended, so progressing 8.9.0, 8.9.1, 8.9.2, 10.0.0, 11.0.0, 12.0.0, 13.0.0, .... The numbers starting 10 or above wouldn't be semantic versions, just incremented when necessary. But 8.9 or earlier nodes could parse them and would understand that they were greater than that node's version.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+     * This is the highest id associated with Ml config versions
+     * that use the legacy, semantic style, representation of the version.
+     */
+    private static final MlConfigVersion MAXIMUM_LEGACY_VERSION = V_8_9_0;
```
We're not going to merge this change to 8.9.1, so it would be better to make this constant 8.10.0 and name it `CUTOVER_VERSION` or something like that. Then the tests based on it should be `before` rather than `onOrBefore`.

edsavage commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,389 @@
+
+    public static final String ML_CONFIG_VERSION_NODE_ATTR = "ml.config_version";
+
+    public static final MlConfigVersion ZERO = registerLegacyMlConfigVersion(0, "00000000-0000-0000-0000-000000000000");
```
Thanks for the heads up @thecoop, we'll keep an eye on it.

edsavage committed:
Merged main

edsavage committed:
Implement the VersionId interface

Future proof the MlConfigVersion class by implementing the VersionId
interface.

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into ml_config_version

edsavage committed:
Attend to review comments

* Don't use MlConfigVersion for versioning index mappings
* Define id of MlConfigVersion to be 10000000 + version. e.g. V_10 has id 10000010, V_11 has id 10000011 etc.
* Print MlConfigVersion numbers in the same style as Version e.g V_10 prints ats 10.0.0, V_11 as 11.0.0 etc.

edsavage committed:
Remove unnecessary overrides

edsavage committed:
Adjust tests to reflect review changes

Account for the fact that the ML index mappings no longer use MlConfigVersion

edsavage committed:
Further clean up of indices mappings

edsavage committed:
Fix failing integration test

edsavage committed:
Merge branch 'main' of github.com:elasticsearch/elasticsearch into ml_config_version

edsavage committed:
Compilation fix

edsavage committed:
Update comment

edsavage committed:
Re-add removed test case

edsavage committed:
Attend to review comments

edsavage committed:
Fix failing test case

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    public static void writeVersion(MlConfigVersion version, StreamOutput out) throws IOException {
+        out.writeVInt(version.id);
```
```suggestion
        // For version 8.10.0 and earlier this must be binary compatible with Version.readVersion()
        out.writeVInt(version.id);
```

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    public static MlConfigVersion readVersion(StreamInput in) throws IOException {
+        return fromId(in.readVInt());
```
```suggestion
        // For version 8.10.0 and earlier this must be binary compatible with Version.writeVersion()
        return fromId(in.readVInt());
```

droberts195 reviewed (commented):
I think if you make this last round of suggested changes and get the tests to work with them then this should be ready to merge.

The nasty scenario to consider is the case of upgrading from one pre-release 8.10.0 snapshot that doesn't contain this functionality to another pre-release 8.10.0 snapshot that does have it. And I don't think the BWC tests in this repo can test that unfortunately.

Even after all these changes I've realised there's still one more BWC problem. If we serialize a new style version to an old node over the wire, then it will be received as a `Version` object. That can cope with major versions up to 99. So if we get through 89 ML config version changes before the real version 10.0.0 is released then we'll have a problem. I think we'll get away with it based on the rate of config format changes to date.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,343 @@
+     * Detached ML config versions added below here.
+     */
+
+    public static final MlConfigVersion V_10 = registerMlConfigVersion(10_00_00_10, "4B940FD9-BEDD-4589-8E08-02D9B480B22D");
```
Actually, I was thinking more about BWC with old nodes, and it would be best if the `Version` class saw the same view whether parsing the printed output of this class or receiving the `id` over the wire.

```suggestion
    public static final MlConfigVersion V_10 = registerMlConfigVersion(10_00_00_99, "4B940FD9-BEDD-4589-8E08-02D9B480B22D");
```

And `V_11` should be `11_00_00_99`, `V_12` should be `12_00_00_99`, etc.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,343 @@
+ * A new ML config version should be added <em>every time</em> a change is made to the serialization protocol of one or more classes.
+ * Each ML config version should only be used in a single merged commit (apart from BwC versions copied from {@link Version}).
+ * <p>
+ * To add a new ML config version, add a new constant at the bottom of the list that is one greater than the current highest version,
```
```suggestion
 * To add a new ML config version, add a new constant at the bottom of the list that is one million greater than the current highest version,
```

edsavage committed:
Attend to further review comments

edsavage commented:
retest

droberts195 reviewed (commented):
This is getting very close now. I just saw a couple of things left that could cause problems in mixed version clusters.

edsavage commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    public static MlConfigVersion fromNode(DiscoveryNode node) {
+        return fromString(node.getAttributes().get(ML_CONFIG_VERSION_NODE_ATTR));
```
> Is there a reason not to accept my suggestion here? This method as it stands will return `fromString(null)` in several different situations, which is undesirable.

Sorry, my mistake!

edsavage committed:
Attend to further review comments

droberts195 reviewed (approved):
LGTM if CI passes 🤞
